### PR TITLE
Added better support for formats that store triples.

### PR DESCRIPTION
**Pull request by jessevdam**
We use semantic data in our pipelines, however for the tools we use it does not matter in which format it is actually stored. So to be able to define 'triples' in the tools we created a format base class 'triples' with the following sub classes 'N-Triples', 'N3','Turtle','RDF/XML' and 'JSON-LD'. To ease the uploading I included a few simple sniffers to auto detect the format. 

We moved the already present 'RDF/XML' format from the graph.py to the triples.py file and sub classed it from the triples base class.

### CODE REVIEW

**Comment by carlfeberhard** [comment]
+1 
You should remove the \* import, but otherwise looks good to me.

**Comment by jmchilton** [comment]
Thanks @jessevdam - these look great. I have opened a PR against your repo here https://github.com/jessevdam/galaxy/pull/1 that I think addresses a few smaller concerns I have. The PR
- Includes PEP-8 fixes for the file to ensure compliance with Galaxy coding standards. e097de3570cb525687ee1800c1582375c6e22b94
- Restores as deprecated older reference to Rdf - so people with older `datatype_conf.xml` files will be warned of the problem but the datatype won't disappear. e714759d409a3099ba04284de8167f66af32424e
- Uses safer file handling in the sniffers (introducing generic infrastructure for doing this that the rest of the datatypes should be retro-fitted to use as well someday). 7038c90b19e5692fba011f6d37e54f4469249fa5

**Comment by jessevdam** [comment]
I tested the sniffers and found a few small errors, which I fixed.

File: lib/galaxy/datatypes/data.py [hunk 1]
```diff
@@ -864,6 +884,23 @@ def regex_line_dataprovider( self, dataset, **settings ):
...
+            start_of_file = f.read(size)
+            lines = start_of_file.splitlines()
+            return lines[0:num_lines]
+
+    def safe_readlines(self, path, num_lines=1, size=5 * 1024):
+        return "".join( self.__safe_readlines( path, num_lines, size ) )
+
+    def safe_readline(self, path):
+        return self.__safe_readlines(path)[0]
+
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/data.py]: As mentioned in https://github.com/jessevdam/galaxy/pull/1#discussion_r33151981, according to https://docs.python.org/2/library/stdtypes.html#file.readline , `readline()`/`readlines()` already have a `size`/`sizehint` parameter respectively which should be enough to be "safe".

Comment (jessevdam) [inline lib/galaxy/datatypes/data.py]: Sorry I missed your message, I will fix this tomorrow and use the file.readline function with the size parameter.

Comment (jessevdam) [inline lib/galaxy/datatypes/data.py]: I changed it, now it uses the size / sizehint parameter from the readline function

**Comment by jessevdam** [comment]
I included the comment from nsoranzo and made the CL build pass again, are there any other items that I have to do or include?

**Comment by yhoogstrate** [comment]
I think some sniffers can be a bit more specific. Example:

```
<?xml version="1.0"?>
<!DOCTYPE xbel PUBLIC 
       "+//IDN python.org//DTD XML Bookmark Exchange
        Language 1.0//EN//XML" 
       "http://www.python.org/topics/xml/dtds/xbel-1.0.dtd">
<xbel version="1.0">
   <bookmark href=
"http://www.w3.org/1999/02/22-rdf-syntax-ns#">
      <title>WC3 RDF syntax</title>
      <desc>
      </desc>
   </bookmark>
</xbel>
```

The RDF sniffer gives a `True` for this valid XBEL file.

**Comment by nsoranzo** [comment]
@jessevdam Thanks a lot for these fixes!

File: lib/galaxy/datatypes/triples.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    """
+    The N-Triples triple data format
+    """
+    edam_format = "format_3256"
+    file_ext = "nt"
+
+    def sniff( self, filename ):
+        with open(filename, "r") as f:
+            # <http://example.org/dir/relfile> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.org/type> .
+            if re.compile( r'<[^>]*>\s<[^>]*>\s<[^>]*>\s\.' ).search( f.readline( 1024 ) ):
```
Comment (bgruening) [inline lib/galaxy/datatypes/triples.py:46]: `re.match(pattern, string)` is for one-time consumption faster I think.

**Comment by bgruening** [comment]
@yhoogstrate this is a mean example ;) But you are absolutely right!
I think we could check if `<[RDFrdf].*>` exists in addition.

**Comment by jessevdam** [comment]
@yhoogstrate @bgruening made small addition to make RDF sniffer a bit more specific. However it was a mean example and it is not that important that it always goes correct, users can specify or change if it goes wrong.

**Comment by jessevdam** [comment]
@bgruening we need the search function as I need to find the regexp somewhere in the string and when it finds one results is stops searching.

**Comment by bgruening** [comment]
> @yhoogstrate @bgruening made small addition to make RDF sniffer a bit more specific. However it was a mean example and it is not that important that it always goes correct, users can specify or change if it goes wrong.

I tend to disagree here. This is important and we need to make sniffers as failure proof as possible. But your adjustments looks good to me.

> @bgruening we need the search function as I need to find the regexp somewhere in the string and when it finds one results is stops searching.

You can also use `re.search(pattern, string)`. My point was that `compile()` is in this position an overhead. Moving it at the top, as class property might make sense.

**Comment by yhoogstrate** [comment]
Sorry for the mean example. The point I wanted to make was that the regex is not explicitly looking for RDF tags within XML encapsulated tags.

What you have changed is redundant, I think. If the regex returns `True`, the second comparison is automatically also `True` because 'RDF' is a substring of 'HTTP://WWW.W3.ORG/1999/02/22-RDF-SYNTAX-NS#'. I would change the second match to '<rdf:RDF' and I would also not use the upper(), because the WC3 validator marks files as invalid if '<rfd:' is not in lowercase.  You would then get something like:

`if re.compile( r'xmlns:[^=]*="http://www.w3.org/1999/02/22-rdf-syntax-ns#"' ).search( firstlines ) and "<rdf:RDF" in firstlines:`

Thanks for the good work!!

**Comment by jessevdam** [comment]
I merged the last changed and improved the RDF/XML sniffer to handle the example given by yhoogstrate.

File: lib/galaxy/datatypes/data.py [hunk 3]
```diff
@@ -46,6 +46,26 @@ def __init__( cls, name, bases, dict_ ):
...
+    """
+    old_init = klass.__init__
+
+    def new_init(self, *args, **kwargs):
+        old_location = "%s.%s" % (klass.__module__, klass.__name__)
+        new_klass = klass.__bases__[0]
+        new_location = "%s.%s" % (new_klass.__module__, new_klass.__name__)
+        message = "Using deprecated reference to class [%s] " % old_location
+        message += "update datatype configuration to reference [%s]" % new_location
+        log.warn(message)
```
Comment (blankenberg) [inline lib/galaxy/datatypes/data.py]: This will write to stderr when setting metadata, causing jobs to fail.

**Comment by martenson** [comment]
Hi @jessevdam, please remove the warn pointed out above and I will merge your request. I am sorry this is taking so long. :/

**Comment by jessevdam** [comment]
@martenson changed the log.warn to a print statement, let me know if this is good otherwise I will remove the print statement completely

File: lib/galaxy/datatypes/data.py [hunk 4]
```diff
@@ -46,6 +46,26 @@ def __init__( cls, name, bases, dict_ ):
         metadata.Statement.process( cls )
 
 
+def deprecated_datatype_reference(klass):
```
Comment (martenson) [inline lib/galaxy/datatypes/data.py]: this (and below) should probably be `class`

**Comment by martenson** [comment]
Looks good to me. Could you please review @blankenberg ?

**Comment by jmchilton** [comment]
Ugh - I've thought a lot about this and gone back and forth and because the way visualization references datatypes - this breaks backward compatibility. Can we just move all of `triples.py` into `graph.py` and update the datatype definitions? Is that acceptable or are we just creating further future backward compatibility problems when we have to give them a name that makes sense.

Also the print is probably still a problem, I would just eliminate it all together if we still want to keep these in triples.py we should just make it disruptive and move things outright all at once.

Very sorry for being indecisive and leading you astray a few times :(.

**Comment by jessevdam** [comment]
For me it is not a problem to merge triples.py back into graph.py and remove the print statement again. But I do not understand with what you means with "Is that acceptable or are we just creating further future backward compatibility problems when we have to give them a name that makes sense." Do you like to change the names of the classes in triples.py?

**Comment by jmchilton** [comment]
Thanks for your patience @jessevdam and sorry for all the delays, we will try to do better next time. These seem like really nice enhancements to really interesting datatypes.

**Comment by jessevdam** [comment]
Thanks for including the patch as we really needed it :P

**Comment by martenson** [comment]
thank you @jessevdam !

**Pull request merged**